File: java-service/src/main/java/de/java/entities/Article.java
package de.java.entities;

/**
 * Created with IntelliJ IDEA.
 * User: christian
 * Date: 11/20/12
 * Time: 7:25 PM
 * To change this template use File | Settings | File Templates.
 */
public class Article extends AbstractEntity {

	private String title;
	private String content;
	private Category category;
	private Tag tag;

	public Article(Long id, Category category, String content, String title) {
		this(id, category, content, title, null);
	}

	public Article(Long id, Category category, String content, String title, Tag tag) {
		super(id);
		this.category = category;
		this.content = content;
		this.title = title;
		this.tag = tag;
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public String getContent() {
		return content;
	}

	public void setContent(String content) {
		this.content = content;
	}

	public Category getCategory() {
		return category;
	}

	public void setCategory(Category category) {
		this.category = category;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Article)) return false;
		if (!super.equals(o)) return false;

		Article article = (Article) o;

		if (category != null ? !category.equals(article.category) : article.category != null) return false;
		if (content != null ? !content.equals(article.content) : article.content != null) return false;
		if (tag != null ? !tag.equals(article.tag) : article.tag != null) return false;
		if (title != null ? !title.equals(article.title) : article.title != null) return false;

		return true;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		result = 31 * result + (title != null ? title.hashCode() : 0);
		result = 31 * result + (content != null ? content.hashCode() : 0);
		result = 31 * result + (category != null ? category.hashCode() : 0);
		result = 31 * result + (tag != null ? tag.hashCode() : 0);
		return result;
	}

	@Override
	public String toString() {
		return "Article{" +
				"category=" + category +
				", title='" + title + '\'' +
				", content='" + content + '\'' +
				", tag=" + tag +
				"} " + super.toString();
	}
}

File: lift-client/src/main/scala/code/snippet/Article.scala
package code.snippet

import net.liftweb._
import common._
import http._
import sitemap._
import util._
import mapper._
import Helpers._
import de.java.services.BlogJavaService
import xml.NodeSeq


/**
 * Menu.param[Long] is not possible -> using a case class instead
 * @param id knows the article id
 */
case class ArticleId(id: Long)

/**
 *
 * Created with IntelliJ IDEA.
 * User: christian
 * Date: 11/20/12
 * Time: 5:31 PM
 * To change this template use File | Settings | File Templates.
 */
object Article {

	lazy val menu = Menu.param[ArticleId]("Article",
																	 		  "Reading an article",
	                                      idAsString => getBoxIncludingId(idAsString),
	                                      articleId => getIdForArticle(articleId)
																	 		 ) / "blog" / "article" / *


	def getBoxIncludingId(articleId: String) : Box[ArticleId] = {
		try {
			val box = new Full[ArticleId](ArticleId(articleId.toLong))
			println("Returning a Full box including the article id=" + box)
			box
		} catch {
				case e : Exception => {
					println("Something went wrong while trying to convert given id=" + articleId)
					return new Full[ArticleId](ArticleId(-1)) // <--- there should be a better solution for it ?!
				}
		}
	}

	def getIdForArticle(articleId: ArticleId) : String = {
		val id = articleId.id.toString
		println("Returning requested article id=" + id)
		id
	}
}


/**
 * Created with IntelliJ IDEA.
 * User: christian
 * Date: 11/20/12
 * Time: 5:31 PM
 * To change this template use File | Settings | File Templates.
 */
class Article(articleId: ArticleId) extends CommonSnippet {

	def render() : CssSel = {
		val article = BlogJavaService.getArticle(articleId.id)

		// in real world exception handling should be the better choice
		if (article != null) {
			return "h1 *" #> article.getTitle & "@article-content" #> article.getContent
		}
		println("article not found! Given id=" + articleId.id)
		S.redirectTo("/blog")
	}

}
